pkg/router/controller/factory/factory.go

```diff
@@ -49,26 +82,133 @@ func (factory *RouterControllerFactory) Create(plugin router.Plugin) *controller
 	}
 }
 
+// CreateNotifier begins listing and watching against the API server for the desired route and endpoint
+// resources. It spawns child goroutines that cannot be terminated. It is a more efficient store of a
+// route system.
+func (factory *RouterControllerFactory) CreateNotifier(changed func()) RoutesByHost {
+	keyFn := cache.MetaNamespaceKeyFunc
+	routeStore := cache.NewIndexer(keyFn, cache.Indexers{"host": hostIndexFunc})
+	routeEventQueue := oscache.NewEventQueueForStore(keyFn, routeStore)
+	cache.NewReflector(&routeLW{
+		client:    factory.OSClient,
+		namespace: factory.Namespace,
+		field:     factory.Fields,
+		label:     factory.Labels,
+	}, &routeapi.Route{}, routeEventQueue, factory.ResyncInterval).Run()
+
+	endpointStore := cache.NewStore(keyFn)
+	endpointsEventQueue := oscache.NewEventQueueForStore(keyFn, endpointStore)
+	cache.NewReflector(&endpointsLW{
+		client:    factory.KClient,
+		namespace: factory.Namespace,
+		// we do not scope endpoints by labels or fields because the route labels != endpoints labels
+	}, &kapi.Endpoints{}, endpointsEventQueue, factory.ResyncInterval).Run()
+
+	go util.Until(func() {
+		for {
+			if _, _, err := routeEventQueue.Pop(); err != nil {
+				return
+			}
+			changed()
+		}
+	}, time.Second, util.NeverStop)
+	go util.Until(func() {
+		for {
+			if _, _, err := endpointsEventQueue.Pop(); err != nil {
+				return
+			}
+			changed()
```

Comment:
Yeah I'm playing around with this for a future index based store (where the
template router gets a lot simpler).  I want to queue up changes to the
config over several seconds before applying them, and also get better at
identifying when things actually change.  However it's not ready yet.

On Aug 31, 2015, at 7:49 PM, Ram Ranganathan <notifications@github.com>
wrote:

In pkg/router/controller/factory/factory.go
<https://github.com/openshift/origin/pull/4428#discussion_r38373277>:

> +	}, &kapi.Endpoints{}, endpointsEventQueue, factory.ResyncInterval).Run()
> +
> +	go util.Until(func() {
> +		for {
> +			if _, _, err := routeEventQueue.Pop(); err != nil {
> +				return
> +			}
> +			changed()
> +		}
> +	}, time.Second, util.NeverStop)
> +	go util.Until(func() {
> +		for {
> +			if _, _, err := endpointsEventQueue.Pop(); err != nil {
> +				return
> +			}
> +			changed()

Would be good to add a parameter to the callback changed to indicate what
changed - endpoint/route.
This doesn't seem to be used currently, right?

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4428/files#r38373277>.